services/port_manager/pom.xml

```diff
@@ -41,6 +41,12 @@
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.junit.vintage</groupId>
+                    <artifactId>junit-vintage-engine</artifactId>
```

Comment:
@chenpiaoping What does this exclusions block do? It could skip the portion of UTs which potentially fails if enabled, right? Let us fix the actual problem directly.

Reply:
Yes, If we delete it, Mockito won't work, but the reason is unknown. I suspect it conflicts with Mockito.

Reply:
Not sure about this. Other services also use Mockito without this exclusion block, and those UTs work okay, for example, vpc/subnet managers.

Reply:
Well, I'll try to find the real reason.

Reply:
4.29.6. Testing
While it is possible to use JUnit 4 to test Kotlin code, JUnit 5 is provided by default and is recommended. JUnit 5 enables a test class to be instantiated once and reused for all of the class’s tests. This makes it possible to use @BeforeClass and @AfterClass annotations on non-static methods, which is a good fit for Kotlin.

JUnit 5 is the default and the vintage engine is provided for backward compatibility with JUnit 4. If you don’t use it, exclude org.junit.vintange:junit-vintage-engine. You also need to switch test instance lifecycle to "per-class".

To mock Kotlin classes, MockK is recommended. If you need the Mockk equivalent of the Mockito specific @MockBean and @SpyBean annotations, you can use SpringMockK which provides similar @MockkBean and @SpykBean annotations.


seens

Reply:
It seems that the combination of Junit5 and Mockito is not very friendly, so if we need to use Mockito, we'd better use Junit4, so we need junit-vintage-engine to support Junit4? Do I understand correctly?

Reply:
@xieus @kimeunju108

Reply:
Junit 5 requires Java 8 or later, but I found vpc/subnet manager use  java1.7 in pom.xml:
<maven.compiler.source>1.7</maven.compiler.source>
<maven.compiler.target>1.7</maven.compiler.target>
and use Junit4 in UTs:
import org.junit.Test;
but port manager requires Java 8 or later(to support lambda)

Reply:
@chenpiaoping By "MockK", do you mean "Mockito"?  Mixing multiple JUnit may not be a good idea as it could cause a bit of headache for compatibility. 

JUnit5 + Mockito2 is possible as suggested here: https://dzone.com/articles/spring-boot-2-with-junit-5-and-mockito-2-for-unit

Maybe worth trying?

Reply:
Regarding java version, this is a good catch. We should at least use java 8 in vpc/subnet manager at the first place. @kevin-zhonghao

Reply:
> @chenpiaoping By "MockK", do you mean "Mockito"? Mixing multiple JUnit may not be a good idea as it could cause a bit of headache for compatibility.
> 
> JUnit5 + Mockito2 is possible as suggested here: https://dzone.com/articles/spring-boot-2-with-junit-5-and-mockito-2-for-unit
> 
> Maybe worth trying?

yes, I tried that, but it didn't work.

Reply:
spring-boot-starter-test actually contains Junit5 and Junit4 (by including junit-vintage-engine), here is its xml：
<?xml version="1.0" encoding="UTF-8"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <parent>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starters</artifactId>
    <version>2.2.6.RELEASE</version>
  </parent>
  <groupId>org.springframework.boot</groupId>
  <artifactId>spring-boot-starter-test</artifactId>
  <version>2.2.6.RELEASE</version>
  <name>Spring Boot Test Starter</name>
  <description>Starter for testing Spring Boot applications with libraries including
		JUnit, Hamcrest and Mockito</description>
  <url>https://projects.spring.io/spring-boot/#/spring-boot-parent/spring-boot-starters/spring-boot-starter-test</url>
  <organization>
    <name>Pivotal Software, Inc.</name>
    <url>https://spring.io</url>
  </organization>
  <licenses>
    <license>
      <name>Apache License, Version 2.0</name>
      <url>https://www.apache.org/licenses/LICENSE-2.0</url>
    </license>
  </licenses>
  <developers>
    <developer>
      <name>Pivotal</name>
      <email>info@pivotal.io</email>
      <organization>Pivotal Software, Inc.</organization>
      <organizationUrl>https://www.spring.io</organizationUrl>
    </developer>
  </developers>
  <scm>
    <connection>scm:git:git://github.com/spring-projects/spring-boot.git</connection>
    <developerConnection>scm:git:ssh://git@github.com/spring-projects/spring-boot.git</developerConnection>
    <url>https://github.com/spring-projects/spring-boot</url>
  </scm>
  <issueManagement>
    <system>Github</system>
    <url>https://github.com/spring-projects/spring-boot/issues</url>
  </issueManagement>
  <dependencies>
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter</artifactId>
      <version>2.2.6.RELEASE</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-test</artifactId>
      <version>2.2.6.RELEASE</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-test-autoconfigure</artifactId>
      <version>2.2.6.RELEASE</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>com.jayway.jsonpath</groupId>
      <artifactId>json-path</artifactId>
      <version>2.4.0</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>jakarta.xml.bind</groupId>
      <artifactId>jakarta.xml.bind-api</artifactId>
      <version>2.3.3</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter</artifactId>
      <version>5.5.2</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.junit.vintage</groupId>
      <artifactId>junit-vintage-engine</artifactId>
      <version>5.5.2</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <artifactId>hamcrest-core</artifactId>
          <groupId>org.hamcrest</groupId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.mockito</groupId>
      <artifactId>mockito-junit-jupiter</artifactId>
      <version>3.1.0</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.assertj</groupId>
      <artifactId>assertj-core</artifactId>
      <version>3.13.2</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.hamcrest</groupId>
      <artifactId>hamcrest</artifactId>
      <version>2.1</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.mockito</groupId>
      <artifactId>mockito-core</artifactId>
      <version>3.1.0</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.skyscreamer</groupId>
      <artifactId>jsonassert</artifactId>
      <version>1.5.0</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.springframework</groupId>
      <artifactId>spring-core</artifactId>
      <version>5.2.5.RELEASE</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.springframework</groupId>
      <artifactId>spring-test</artifactId>
      <version>5.2.5.RELEASE</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.xmlunit</groupId>
      <artifactId>xmlunit-core</artifactId>
      <version>2.6.4</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <artifactId>jaxb-api</artifactId>
          <groupId>javax.xml.bind</groupId>
        </exclusion>
      </exclusions>
    </dependency>
  </dependencies>
</project>

Reply:
with junit-vintage-engine maybe ok? I found some examples like this on the Internet.

Reply:
Since the scope of the built-in junit-vintage-engine is compile, When I change it to test, Mockito is OK, but we don't want to modify the pom.xml file of springboot, so we include junit-vintage-engine through exclusions. The question is why Mockito doesn't work when scope is compile.

Reply:
The starter also brings the vintage engine so that you can run both JUnit 4 and JUnit 5 tests. If you have migrated your tests to JUnit 5, you should exclude JUnit 4 support, as shown in the following example:<dependency>     <groupId>org.springframework.boot</groupId>     <artifactId>spring-boot-starter-test</artifactId>     <scope>test</scope>     <exclusions>         <exclusion>             <groupId>org.junit.vintage</groupId>             <artifactId>junit-vintage-engine</artifactId>         </exclusion>     </exclusions> </dependency>
--


https://docs.spring.io/spring-boot/docs/2.2.7.RELEASE/reference/html/spring-boot-features.html#boot-features-testing

Reply:
The starter also brings the vintage engine so that you can run both JUnit 4 and JUnit 5 tests. If you have migrated your tests to JUnit 5, you should exclude JUnit 4 support, as shown in the following example:
<dependency>     <groupId>org.springframework.boot</groupId>     <artifactId>spring-boot-starter-test</artifactId>     <scope>test</scope>     <exclusions>         <exclusion>             <groupId>org.junit.vintage</groupId>             <artifactId>junit-vintage-engine</artifactId>         </exclusion>     </exclusions> </dependency>

Reply:
https://docs.spring.io/spring-boot/docs/2.2.7.RELEASE/reference/html/spring-boot-features.html#boot-features-testing

The starter also brings the vintage engine so that you can run both JUnit 4 and JUnit 5 tests. If you have migrated your tests to JUnit 5, you should exclude JUnit 4 support, as shown in the following example:
--

Reply:
<dependency>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-test</artifactId>
    <scope>test</scope>
    <exclusions>
        <exclusion>
            <groupId>org.junit.vintage</groupId>
            <artifactId>junit-vintage-engine</artifactId>
        </exclusion>
    </exclusions>
</dependency>

Reply:
mock service can work, that is the reason why vpc/subnet manager is ok.

Reply:
@chenpiaoping If you click https://github.com/futurewei-cloud/alcor/pull/197/checks?check_run_id=684996404, you could find the current merge results shows as follows:

[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 5.15 s - in com.futurewei.alcor.portmanager.PortManagerApplicationTests
2020-05-18 11:04:16.840  INFO 3974 --- [extShutdownHook] o.s.s.concurrent.ThreadPoolTaskExecutor  : Shutting down ExecutorService 'applicationTaskExecutor'
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0

The goal for us is to enable on UTs for every microservice. Port Manager is so important that we definite need it on.

Reply:
Oh, let me check, is there a problem with these outputs?

Reply:
Only one test case is on, which is PortManagerApplicationTests. All the good test cases in PortControllerTest.java are not actually running. That is why I am trying to turn them all on.

Reply:
I can see outputs of all UTs from section Build with Maven.
The snippet:
      HTTP Method = POST
      Request URI = /project/3dda2801-d675-4688-a63f-dcda8d327f50/network_segment_ranges
       Parameters = {}
          Headers = [Content-Type:"application/json;charset=UTF-8"]
             Body = {"network_segment_range":{"project_id":"3dda2801-d675-4688-a63f-dcda8d327f50","id":"9192a4d4-ffff-4ece-b3f0-8d36e3d86000","name":"test_segment","description":"","network_type":"vlan"}}
    Session Attrs = {}

Handler:
             Type = com.futurewei.alcor.vpcmanager.controller.SegmentRangeController

Reply:
That is VPC manager UT.

Reply:
sorry, I read it wrong.

Reply:
Just checking in to see if you are still blocked on this.

Reply:
I'm already on security group. back soon.

Reply:
It is okay. Let us know once it is ready to review then @chenpiaoping

Reply:
ok

Reply:
The problem has been solved.

Reply:
Cool thanks! What was the catch?

Reply:
org.junit.Test belongs to junit4, after excluding junit4, we should not use it any more, or the UTs would not be executed at compile time.